tiledb/api/src/datatype/arrow.rs

```diff
@@ -46,169 +53,667 @@ impl ArrowPrimitiveTypeNative for f64 {
+            };
+            if item.is_nullable() {
+                // no way to represent null values within a cell
+                Res::None
+            } else if item.data_type().primitive_width().is_none() {
+                /*
+                 * probably there are some cases we can handle,
+                 * but let's omit for now
+                 */
+                Res::None
+            } else if let Some(exact_datatype) = item
+                .metadata()
+                .get(ARROW_FIELD_METADATA_KEY_TILEDB_EXACT_MATCH)
+                .and_then(|s| Datatype::from_string(s))
+            {
+                Res::Exact(exact_datatype, CellValNum::Fixed(len))
+            } else {
+                match from_arrow(item.data_type()) {
+                    Res::None => Res::None,
+                    Res::Inexact(item, item_cell_val) => {
+                        let cell_val_num = match item_cell_val {
+                            CellValNum::Fixed(nz) => {
+                                match nz.checked_mul(len) {
+                                    None => return Res::None,
+                                    Some(nz) => CellValNum::Fixed(nz),
+                                }
+                            }
+                            CellValNum::Var => CellValNum::Var,
+                        };
+                        Res::Inexact(item, cell_val_num)
+                    }
+                    Res::Exact(item, item_cell_val) => {
+                        let cell_val_num = match item_cell_val {
+                            CellValNum::Fixed(nz) => {
+                                match nz.checked_mul(len) {
+                                    None => return Res::None,
+                                    Some(nz) => CellValNum::Fixed(nz),
+                                }
+                            }
+                            CellValNum::Var => CellValNum::Var,
+                        };
+                        Res::Exact(item, cell_val_num)
+                    }
+                }
+            }
+        }
+        ADT::LargeUtf8 => {
+            /*
+             * NB: arrow checks for valid UTF-8 but tiledb does not.
+             * This is not an exact conversion for that reason
+             * because we cannot guarantee invertibility.
+             */
+            Res::Inexact(Datatype::StringUtf8, CellValNum::Var)
+        }
+        ADT::LargeList(ref item) => {
+            if item.is_nullable() {
+                // no way to represent null values within a cell
+                Res::None
+            } else if item.data_type().primitive_width().is_none() {
+                /*
+                 * probably there are some cases we can handle,
+                 * but let's omit for now
+                 */
+                Res::None
+            } else if let Some(exact_datatype) = item
+                .metadata()
+                .get(ARROW_FIELD_METADATA_KEY_TILEDB_EXACT_MATCH)
```

Comment:
I think you're doing what I'm thinking should be done, but calling this `EXACT_MATCH` is really not doing wonderful things to my brain since I consider this the precise opposite of an exact match. According to other code a TileDB Datatype::Int32 is an exact match to Arrow's Int32 type. An in-exact type would be something like "we gave arrow a Datatype::Int64 instead of a Datatype::TimeFemtosecond, so use Femtosecond instead".

We only need the extra metadata for the in-exact case so calling the extra metadata `EXACT_MATCH` strikes me as not at all correct.

Also, I'm suddenly wondering how often we'll even see this in real life? I guess joins and/or something like `insert into foo select from bar` type stuff?

Reply:
I added this because of some weirdness with the `StringAscii` datatype.

Let's say we wanted to do `CREATE new_table AS SELECT * FROM old_table` or something like that, preserving dimensions of the source table.  If the source table has a `StringAscii` dimension then it treats domain, extent and some other things very differently than if we used `UInt8`.

I neglected to write down how exactly this sort of thing affected the testing, and unfortunately I've forgotten... but it did, and doing this resolved it

Reply:
Sure. The TileDB <=> Arrow round tripping is lossy due to type mismatches. Adding type hints to aid in round trip fidelity makes perfect sense. I just think the naming is wrong since we only need the hint in the absence of an exact match.